src/librustc_mir/transform/simplify_cfg.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use repr::*;
+use rustc::middle::const_eval::ConstVal;
+use std::mem;
+use transform::util;
+use transform::MIRPass;
+
+pub struct SimplifyCFG;
+
+impl SimplifyCFG {
+    fn remove_dead_blocks(&self, mir: &mut Mir) {
+        let mut seen: Vec<_> = mir.basic_blocks.iter().map(|_| false).collect();
+
+        // These blocks are always required.
+        seen[START_BLOCK.index()] = true;
+        seen[END_BLOCK.index()] = true;
+        seen[DIVERGE_BLOCK.index()] = true;
+
+        let mut worklist = vec![START_BLOCK];
+        while let Some(bb) = worklist.pop() {
+            for succ in mir.basic_block_data(bb).terminator.successors() {
+                if !seen[succ.index()] {
+                    seen[succ.index()] = true;
+                    worklist.push(*succ);
+                }
+            }
+        }
+
+        util::remove_basic_blocks(mir, &seen);
+    }
+
+    fn collapse_jump_chains(&self, mir: &mut Mir) {
+
+        // Find the target at the end of the jump chain, return None if there is a loop
+        fn final_target(mir: &Mir, mut target: BasicBlock) -> Option<BasicBlock> {
+            // Keep track of already seen blocks to detect loops
+            let mut seen: Vec<BasicBlock> = Vec::with_capacity(8);
+
+            while mir.basic_block_data(target).statements.is_empty() {
+                match mir.basic_block_data(target).terminator {
+                    Terminator::Goto { target: next } |
+                    Terminator::Panic { target: next } => {
```

Comment:
In matching both of these, won't this turn a chain like "goto -> panic" into just "goto"? (e.g. wouldn't that change semantics?)